libs/ebpfnetsh/programs.cpp

```diff
@@ -838,3 +838,88 @@ handle_ebpf_show_programs(
     }
     return status;
 }
+
+template <typename F>
+unsigned long
+handle_ebpf_pinunpin_program_common(
+    LPCWSTR machine, LPWSTR* argv, DWORD current_index, DWORD argc, DWORD flags, LPCVOID data, BOOL* done, F&& f)
+{
+    UNREFERENCED_PARAMETER(machine);
+    UNREFERENCED_PARAMETER(flags);
+    UNREFERENCED_PARAMETER(data);
+    UNREFERENCED_PARAMETER(done);
+
+    TAG_TYPE tags[] = {
+        {TOKEN_ID, NS_REQ_PRESENT, FALSE},
+        {TOKEN_PINPATH, NS_REQ_ZERO, FALSE},
+    };
+    const int ID_INDEX = 0;
+    const int PINPATH_INDEX = 1;
+
+    unsigned long tag_type[_countof(tags)]{};
+
+    unsigned long status =
+        PreprocessCommand(nullptr, argv, current_index, argc, tags, _countof(tags), 0, _countof(tags), tag_type);
+    if (status != EBPF_SUCCESS) {
+        return status;
+    }
+
+    std::string pinpath;
+    uint32_t id = 0;
+
+    for (int i = 0; (status == NO_ERROR) && ((i + current_index) < argc); i++) {
+        switch (tag_type[i]) {
+        case PINPATH_INDEX:
+            pinpath = down_cast_from_wstring(std::wstring(argv[current_index + i]));
+            break;
+
+        case ID_INDEX:
+            id = wcstoul(argv[current_index + i], nullptr, 0);
+            break;
+        }
+    }
+
+    auto fd = bpf_prog_get_fd_by_id(id);
+    if (fd < 0)
+        return ERROR_INVALID_PARAMETER;
+
+    if (pinpath.empty()) {
+        bpf_prog_info info{};
+        uint32_t size = sizeof(info);
+        if (bpf_obj_get_info_by_fd(fd, &info, &size) == 0) {
+            pinpath = info.name;
+        }
+    }
+    if (!pinpath.empty()) {
+        status = f(fd, pinpath.c_str());
+    } else {
+        status = ERROR_INVALID_PARAMETER;
+    }
+    Platform::_close(fd);
+
+    return status;
+}
+
+// The following function uses windows specific type as an input to match
+// definition of "FN_HANDLE_CMD" in public file of NetSh.h
+unsigned long
+handle_ebpf_pin_program(
+    LPCWSTR machine, LPWSTR* argv, DWORD current_index, DWORD argc, DWORD flags, LPCVOID data, BOOL* done)
+{
+    return handle_ebpf_pinunpin_program_common(
+        machine, argv, current_index, argc, flags, data, done, [](auto fd, auto pinpath) {
+            return bpf_obj_pin(fd, pinpath);
+        });
+}
+
+// The following function uses windows specific type as an input to match
+// definition of "FN_HANDLE_CMD" in public file of NetSh.h
+unsigned long
+handle_ebpf_unpin_program(
+    LPCWSTR machine, LPWSTR* argv, DWORD current_index, DWORD argc, DWORD flags, LPCVOID data, BOOL* done)
+{
+    return handle_ebpf_pinunpin_program_common(
+        machine, argv, current_index, argc, flags, data, done, [](auto, auto pinpath) {
+            return ebpf_object_unpin(pinpath);
```

Comment:
Line 308 has a function `_unpin_program_by_id()`.  Do we really need two ways to do an unpin or can this be cleaned up?

Reply:
Can you at least add comments to both about this fact?  Thanks!